compile/api/src/main/scala/sbt/ClassToAPI.scala

```diff
@@ -127,14 +135,48 @@ object ClassToAPI {
   def upperBounds(ts: Array[Type]): api.Type =
     new api.Structure(lzy(types(ts)), lzyEmptyDefArray, lzyEmptyDefArray)
 
-  def fieldToDef(enclPkg: Option[String])(f: Field): api.FieldLike =
+  @deprecated("Use fieldToDef[4] instead", "0.13.9")
+  def fieldToDef(enclPkg: Option[String])(f: Field): api.FieldLike = {
+    val c = f.getDeclaringClass()
+    fieldToDef(c, classFileForClass(c), enclPkg)(f)
+  }
+
+  def fieldToDef(c: Class[_], _cf: => ClassFile, enclPkg: Option[String])(f: Field): api.FieldLike =
     {
       val name = f.getName
       val accs = access(f.getModifiers, enclPkg)
       val mods = modifiers(f.getModifiers)
       val annots = annotations(f.getDeclaredAnnotations)
-      val tpe = reference(returnType(f))
-      if (mods.isFinal) new api.Val(tpe, name, accs, mods, annots) else new api.Var(tpe, name, accs, mods, annots)
+      // generate a more specific type for constant fields
+      val specificTpe: Option[api.Type] =
+        if (mods.isFinal) {
+          val cf = _cf
+          val attributeInfos = cf.fields.find(_.name.exists(_ == name)).toSeq.flatMap(_.attributes)
+          // create a singleton type ending with the ConstantValue of this field. because this type
+          // is purely synthetic, it's fine that the name might contain filename-banned characters.
+          attributeInfos.collectFirst {
+            case ai @ classfile.AttributeInfo(Some("ConstantValue"), _) =>
+              try {
+                c.getName.split("\\.").toSeq :+ (name + "$" + cf.constantValue(ai))
```

Comment:
I think it would be useful to elaborate a bit on why these parts (class name, field name, constant value) are included in the singleton type's path. The thought process of a future incremental compiler hacker would have to go something like this. If the class name changes, but the constant does not, should we recompile? (I assume we don't.) So, does it suffice to include the value? But wait! If you change the constant's value, and not the class name, the change won't be detected, so I should include the reference to the field to encode the dependency. 

How about the type of the field (not captured here), though? Does it matter whether I go from `int` to `long`, as, erm, long as the value doesn't change?